File: fsharp/day2_part2_2016.fs

module Day2

open System.IO

let input = File.ReadAllLines "input.txt"

let keypad1 = [| [|'1'; '2'; '3'|]; [|'4'; '5'; '6'|]; [|'7'; '8'; '9'|] |]
let keypad2 = [| [|'0'; '0'; '1'; '0'; '0'|]; [|'0'; '2'; '3'; '4'; '0'|]; [|'5'; '6'; '7'; '8'; '9'|]; [|'0'; 'A'; 'B'; 'C'; '0'|]; [|'0'; '0'; 'D'; '0'; '0'|] |]

let move (keypad: char[][]) (start: char) (direction: char) =
    let row = keypad |> Array.findIndex (fun x -> Array.contains start x)
    let col = keypad.[row] |> Array.findIndex (fun x -> x = start)
    match direction with
    | 'U' -> if row > 0 && keypad.[row-1].[col] <> '0' then keypad.[row-1].[col] else start
    | 'D' -> if row < keypad.Length - 1 && keypad.[row+1].[col] <> '0' then keypad.[row+1].[col] else start
    | 'L' -> if col > 0 && keypad.[row].[col-1] <> '0' then keypad.[row].[col-1] else start
    | 'R' -> if col < keypad.[row].Length - 1 && keypad.[row].[col+1] <> '0' then keypad.[row].[col+1] else start
    | _ -> start

let getCode (keypad: char[][]) (instructions: string[]) =
    let mutable code = ""
    let mutable current = '5'
    for instruction in instructions do
        for direction in instruction do
            current <- move keypad current direction
        code <- code + string current
    code

let result1 = getCode keypad1 input
let result2 = getCode keypad2 input

printfn "%s" result1
printfn "%s" result2

File: haskell/day2_part2_2016.hs

import Data.Char (digitToInt)
import Data.List (foldl')

type Keypad1 = [[Char]]
type Keypad2 = [[Char]]
type Position = (Int, Int)

keypad1 :: Keypad1
keypad1 = ["123", "456", "789"]

keypad2 :: Keypad2
keypad2 = ["  1  ", " 234 ", "56789", " ABC ", "  D  "]

move :: Keypad1 -> Position -> Char -> Position
move keypad (row, col) 'U' = if row > 0 && keypad !! (row - 1) !! col /= ' ' then (row - 1, col) else (row, col)
move keypad (row, col) 'D' = if row < length keypad - 1 && keypad !! (row + 1) !! col /= ' ' then (row + 1, col) else (row, col)
move keypad (row, col) 'L' = if col > 0 && keypad !! row !! (col - 1) /= ' ' then (row, col - 1) else (row, col)
move keypad (row, col) 'R' = if col < length (head keypad) - 1 && keypad !! row !! (col + 1) /= ' ' then (row, col + 1) else (row, col)
move _ pos _ = pos

findStartPos :: Keypad1 -> Char -> Position
findStartPos keypad start = head [(r, c) | r <- [0..length keypad - 1], c <- [0..length (head keypad) - 1], keypad !! r !! c == start]

findStartPos2 :: Keypad2 -> Char -> Position
findStartPos2 keypad start = head [(r, c) | r <- [0..length keypad - 1], c <- [0..length (head keypad) - 1], keypad !! r !! c == start]

solve :: Keypad1 -> String -> String
solve keypad input =
  let startPos = findStartPos keypad '5'
      code = foldl' (\(pos, acc) line ->
                      let newPos = foldl' (move keypad) pos line
                      in (newPos, acc ++ [keypad !! (fst newPos) !! (snd newPos)])
                    ) (startPos, "") (lines input)
  in snd code

solve2 :: Keypad2 -> String -> String
solve2 keypad input =
  let startPos = findStartPos2 keypad '5'
      code = foldl' (\(pos, acc) line ->
                      let newPos = foldl' (move keypad) pos line
                      in (newPos, acc ++ [keypad !! (fst newPos) !! (snd newPos)])
                    ) (startPos, "") (lines input)
  in snd code

main :: IO ()
main = do
  input <- readFile "input.txt"
  putStrLn $ "Part 1: " ++ solve keypad1 input
  putStrLn $ "Part 2: " ++ solve2 keypad2 input
